File: sdks/kotlin/examples/Preview.kt
// Runs a CAVS update preview between two build directories you already have.
//
//   gradle -q runPreview --args="--old Build_v1 --new Build_v2"
package com.cavs.examples

import com.cavs.sdk.CavsClient
import com.cavs.sdk.model.PreviewRequest
import kotlin.system.exitProcess

fun main(args: Array<String>) {
    val oldPath = valueOf(args, "--old")
    val newPath = valueOf(args, "--new")
    if (oldPath == null || newPath == null) {
        System.err.println("usage: preview --old <dir> --new <dir>")
        exitProcess(2)
    }

    CavsClient.create().use { cavs ->
        val report = cavs.preview(
            PreviewRequest(oldPath = oldPath, newPath = newPath, policy = "balanced"),
        ) { e -> e.phase?.let { System.err.println("  [${e.type}] $it") } }

        println("Recommended route: ${report.recommendedRoute}")
        for (r in report.routes) {
            println("  ${r.name.padEnd(16)} ${r.networkBytes} bytes")
        }
    }
}

private fun valueOf(args: Array<String>, flag: String): String? {
    val i = args.indexOf(flag)
    return if (i >= 0 && i + 1 < args.size) args[i + 1] else null
}

File: sdks/node/examples/preview.ts
// Run a CAVS update preview between two build directories.
//
//   npm run build && node dist/examples/preview.js --old Build_v1 --new Build_v2
import { CavsClient } from "../src/index";

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const oldPath = valueOf(args, "--old");
  const newPath = valueOf(args, "--new");
  if (!oldPath || !newPath) {
    console.error("usage: preview --old <dir> --new <dir>");
    process.exit(2);
  }

  const cavs = new CavsClient();
  try {
    const report = await cavs.preview(
      { oldPath, newPath, policy: "balanced" },
      { onProgress: (e) => e.phase && console.error(`  [${e.type}] ${e.phase}`) },
    );
    console.log(`Recommended route: ${report.recommendedRoute}`);
    for (const r of report.routes) {
      console.log(`  ${r.name.padEnd(16)} ${r.networkBytes} bytes`);
    }
  } finally {
    cavs.close();
  }
}

function valueOf(args: string[], flag: string): string | undefined {
  const i = args.indexOf(flag);
  return i >= 0 ? args[i + 1] : undefined;
}

void main();
